packages/pglite-sync/src/index.ts

```diff
@@ -52,14 +78,56 @@ async function createPlugin(
         }
         await pg.transaction(async (tx) => {
           for (const message of messages) {
-            await applyMessageToTable({
+            shapeId ??= stream.shapeId
+
+            if (isChangeMessage(message)) {
+              await applyMessageToTable({
+                pg: tx,
+                message: message,
+                table: options.table,
+                schema: options.schema,
+                mapColumns: options.mapColumns,
+                primaryKey: options.primaryKey,
+                debug,
+              })
+              lastOffsetAdded = message.offset
+            }
+
+            if (isControlMessage(message)) {
+              switch (message.headers.control) {
+                case 'must-refetch':
+                  if (debug) console.log('clearing and refetching shape')
+                  shapeId = undefined
+                  lastOffsetAdded = undefined
+
+                  // TODO: sync into shadow table and reference count
+                  // for now just clear the whole table
+                  await tx.exec(`TRUNCATE ${options.table};`)
```

Comment:
As you've noted we need to ref count and do proper GC. I don't think we can release a version that just truncates as there may be multiple shapes in the same table. That or we block multiple shapes to the same table for a first release?

I know I mentioned splitting this up into multiple PRs, I no realise that we will have problems if we need to do a realise of PGlite before the full lot is merged if we were to merge this as an interim.

Reply:
I will say that currently if for whatever reason a `must-refetch` message is received it is ignored and the stream is restarted, resulting in a replay of the log and a cascade of errors, so I'm not sure if this is truly "breaking" behaviour.

That being said, I could put the truncate behind the presence of the `shapeKey` as well so we can release an undocumented naive persistence and stack PRs

Reply:
adding: the truncating on `must-refetch` is not really related to the shape stream state persistence, but to the handling of the shape stream protocol in general - even if it is ephemeral it might have to replay the stream and currently that's not handled at all, but I understand if we want to opt for durability in those cases

Reply:
I suppose comes down to how soon we can add GC. We could just release with a nice big warning, and log a message in the console when it happens linking to the docs explains the truncate.

Reply:
It's definitely broken without handling must-refetch e.g. https://discord.com/channels/933657521581858818/933657523049885698/1287880874133553192

Could we just put a lock per table so only one shape can write to the table? That seems like a reasonable restriction for now.

Reply:
Yep, agreed. I'm happy to not have a lock for now if we think we can get GC in quick soon (it should be too hard to build), and then just document with a nice big warning.

Ideally the truncate (or GC when we get it) and then re-insert of all rows should happen **in the same transaction** so that this doesn't trigger multiple live queries where you would get a flash of no content. It could result in some pages even temporarily showing a 404 while the new insert is happening.

Reply:
Until we have GC we could set a `truncateNeeded` flag that triggers a truncate at the start of the next transaction.